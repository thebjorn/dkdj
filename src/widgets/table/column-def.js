
import Class from "../../lifecycle/coldboot/dk-class";
import dk from "../../dk-obj";
import {value as format_value} from "../../data/datacore/dk-format"
import widgetmap from "../../widgetcore/dk-widgetmap";
import {TextInputWidget} from "../../forms/widgets";


export class ColumnDef extends Class {
    constructor(...args) {
        super({
            isa:         'dk.table.ColumnDef',
            name:        '',                        // name of column (eg. for computed cols)
            field:       '',                        // field definition {name:.., pos:.., string:..,...}
            label:       '',                        // text to put in th element
            sortable:    undefined,                 // sortable override
            description: '',                        // description
            align:       undefined,                 // cell alignment
            format:      format_value,  // formatting function
            type:        undefined,                 // data type of column values
            widget_type: TextInputWidget,     // widget to use when editing a cell value
            empty:       '',                        // value to use for empty cells
            table:       undefined,                 // the DataTable (subclass) we belong to
            _alignment_map: {
                int: 'right'
            }
        }, ...args);
        this.name = this.name || this.field;
    }

    alignment() {
        return this.align || this._alignment_map[this.type] || 'left';
    }

    get_value(record) {
        //dk.info("getting value from columndef", record, this, record[this.name]);
        if(record)
            return record[this.name];
    }

    /*
     *  Update self with info from the datasource.
     */
    bind_to_field(field) {
        this.field = field;
        if (!this.label) this.label = field.label || field.name;
        if (!this.type) this.type = field.type;

        let wtype;
        if (field.widget) {
            // wtype = widgetmap.get(field.widget);
            wtype = widgetmap[field.widget];
        } else {
            wtype = widgetmap[this.type];
        }

        if (wtype === undefined) {
            // dk.warn("Unknown widget: ", field.widget, field);
            wtype = TextInputWidget;
        }
        this.widget_type = wtype;
        this.widget_data = field.data;
        this.widget_url = field.url;
    }
}
