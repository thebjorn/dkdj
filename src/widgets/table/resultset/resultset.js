import dk from "../../../dk-obj";
import {Widget} from "../../../widgetcore/dk-widget";
import {SearchWidget} from "../../search-widget";
import {PagerWidget} from "../../pager-widget";
import {dkwarning} from "../../../lifecycle/coldboot/dkwarning";
import {dedent} from "../../../text/template-functions";
        

export class ResultSet extends Widget {
    constructor(...args) {
        super({
            // type: 'resultset',

            filter_size: 20,     // percent

            structure: {
                classes: ['clearfix'],
                rowbx: {
                    css: {
                        display: 'flex',
                        alignItems: 'flex-start'
                    },
                    filterbx: {
                        css: {
                            marginRight: 4,
                            width: 180
                            // flex: '0 0 180px'
                            // float: 'left'
                        }
                    },
                    data: {
                        classes: ['panel', 'panel-default'],
                        css: {
                            flex: 1
                            //float: 'right'
                        },
                        header: {
                            classes: ['panel-heading', 'clearfix'],
                            search: {
                                classes: ['pull-left'],
                                css: {
                                    width: '40%'
                                }
                            },
                            commands: {
                                classes: ['pull-right'],
                                add: {},
                                csv: {}
                            },
                            stats: {
                                template: '<div/>',
                                classes: ['pull-right'],
                                css: {textAlign: 'right', marginRight: '1ex'}
                            }
                        },
                        content: {
                            query: '.result-table',
                            template: '<table/>',
                            classes: ['result-table', 'table', 'table-bordered', 'table-hover', 'table-condensed']
                        },
                        footer:  {
                            template: '<footer/>',
                            classes: ['panel-footer']
                        }
                    }
                }
            },
            template: {
                root: 'section',
                search:         '<div/>',
                csv:   `<button class="btn excelbtn" data-loading-text="Laster ned...">
                            <span><img height="28" alt="csv" src="https://static.datakortet.no/ikn/ax/button_blue_down.png"></span>
                        </button>`,
                add:            '<img class="add-button" alt="add" height="28" src="https://static.datakortet.no/ikn/ax/button_blue_add.png">',
                commands:       '<div class="commands"/>'
            },
            
            command_buttons: {
                // default command parameters
                add: {
                    show: true,
                    url: '/',
                },
                csv: {
                    show: true,
                    filename: 'filename.csv',
                },
            }
            
        }, ...args);

        // XXX: make button/command handling less ugly...?
        const cmd_btns = this.command_buttons;
        const command_buttons = {
            // default command parameters
            add: {
                show: true,
                url: '/',
            },
            csv: {
                show: true,
                filename: 'filename.csv',
            },
        };
        if (cmd_btns.add === false) command_buttons.add.show = false;
        if (cmd_btns.add) command_buttons.add = dk.merge(command_buttons.add, cmd_btns.add);
        if (cmd_btns.csv === false) command_buttons.csv.show = false;
        if (cmd_btns.csv) command_buttons.csv = dk.merge(command_buttons.csv, cmd_btns.csv);
        this.command_buttons = command_buttons;
        if (cmd_btns.csv && cmd_btns.csv.filename && cmd_btns.csv.filename !== this.command_buttons.csv.filename) {
            this.command_buttons.csv.filename = cmd_btns.csv.filename;
        }

        if (!this.command_buttons.add.show || this.add_widget === false) delete this.structure.rowbx.data.header.commands.add;
        if (!this.command_buttons.csv.show || this.download_widget === false) delete this.structure.rowbx.data.header.commands.csv;
        
        if (!this.dataset) dkwarning(dedent`
            You should define .dataset on the Resultset (as opposed to inside the construct_table function):
            
                dk.ResultSet.create_on(.., {
                    dataset: dk.data.DataSet.create({
                        datasource: ...e.g. url...,
                        pagesize: 5,
                        orphans: 4
                    }),
                    construct_table: function (location, downloadwidget, ds) {
                        table_data: ds,
                        ...
                    }
                });
        `);
    }

    collapse_filter(size) {
        // dk.info("COLLAPSE_FILTER:", arguments, this.rowbx.filterbx, this.widget().width());
        // this.rowbx.filterbx.css('width', size.width);
        // dk.info("size:", size);
        // this.rowbx.filterbx.css('margin-right', -1 * size.height + size.width + 4);
        // dk.info("FILTERBX width:", this.rowbx.filterbx.width());
        // this.rowbx.data.css('width', '80%');
    }

    construct_filter(location, dataset) {
        this.filter = null;
        // override this method to attach a filter.  The default version removes the
        // filter box and stretches the data box to full width.
        location.css({
            marginRight: 0,
            paddingRight: 0,
            width: 0
        });
        return null;
    }

    // noinspection JSUnusedLocalSymbols
    construct_table(location, download, ds) {
        if (!this.table) {
            return this._construction_error("you need to specify a construct_table method");
        }
    }

    // Called automatically when structure.header.search have been created.
    // (because.. structure/layout/widget work together so any structure element `foo`,
    // will look for and call a corresponding `widget.construct_foo(location)` method...
    // yes, magic.)
    construct_search(location) {
        return SearchWidget.create_inside(location, {});
    }

    construct_pager(location, state) {
        // console.log("CONSTRUCTING_PAGER", location, state);
        this.pager = null;
        return PagerWidget.create_inside(location, state);
    }

    _construction_error(msg, type) {
        type = type || 'danger';
        this.widget().append(dk.$('<div/>').addClass('alert alert-' + type).html(msg));
    }

    construct() {
        // subwidget handlers
        if (this.dataset) dk.on(this.dataset, 'fetch-info', info => this._update_info(info));
        this.state = {}; //dk.page.hash.substate(this.id);
        this.header = this.rowbx.data.header;
        this.table = this.construct_table(
            this.rowbx.data.content, 
            this.widget('.excel'), 
            this.dataset
        );
        if (!this.table) return;
        if (this.command_buttons.csv.show && !this.table.download) {
            // we should show the csv download button, but construct_table has not initialized it...
            this.widget('.excelbtn').on('click', () => this.table.table_data.fetch_csv(this.command_buttons.csv.filename)); 
        }
        if (this.command_buttons.add.show) {
            this.widget('.add-button').on('click', () => location.assign(this.command_buttons.add.url));
        }
        // console.log("Resultset:construct:end:construct_table");
        
        this.filter = this.construct_filter(this.rowbx.filterbx, this.table.table_data);
        this.pager = this.construct_pager(this.rowbx.data.footer, {});
        this.table.set_pager(this.pager);
    }
    
    _update_info(info) {
        // console.log("ResultSet:handler:_update_info:", info);
        let count = 'many';
        const start_recnum = info.start_recnum || 0;
        const end_recnum = info.end_recnum || 0;
        const filter_count = info.filter_count || 0;
        const totcount = info.totcount || 0;

        if (end_recnum - start_recnum === 0) count = 'none';
        if (end_recnum - start_recnum === 1) count = 'one';

        let infotxt = '<span class="info">';
        const total = `<span class="tot">(totalt: ${totcount})</span>`;
        // console.log("SWITCH:", count, info);

        switch (count) {
            case 'none':
                infotxt += `viser 0 av ${filter_count}`;
                break;
            case 'one':
                infotxt += `post nr. ${start_recnum} av ${filter_count}`;
                break;
            case 'many':
                infotxt += `viser ${start_recnum}&ndash;${end_recnum-1} av ${filter_count}`;
                break;
        }
        infotxt += '</span>';

        this.header.stats.html(infotxt + total);
    }
    
    handlers() {
        // console.info("DECLARING Resultset.handlers()");
        if (this.filter)   dk.on(this.filter, 'collapse-done', () => this.collapse_filter());
        if (!this.dataset) dk.on(this.table.table_data, 'fetch-info', info => this._update_info(info));

        dk.on(this.table.table_data, 'fetch-data-start', () => this.start_busy());
        dk.on(this.table.table_data, 'fetch-data', () => this.end_busy());

        // FIXME: this assumes that this.filter is a panel that has a .datafilter attribute that 
        //        is a widgets/filter/data-filter.js:DataFilter instance
        if (this.filter) {
            dk.on(this.filter.datafilter, 'filter-change', filtervals => this.table.table_data.set_filter(filtervals));
        }
        dk.on(this.table.table_data, 'fetch-page', () => this.pager.draw());
        dk.on(this.header.search, 'search', (terms) => this.table.set_search(terms));
        dk.on(this.header.search, 'clear', () => this.table.set_search());
    }

}
