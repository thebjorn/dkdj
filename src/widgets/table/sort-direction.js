
// var Widget = require('../../widgetcore/dk-widget.js');


import {Widget} from "../../widgetcore/dk-widget";
import {icon} from "../dk-icon-library";

export class SortDirection extends Widget {
    constructor(...args) {
        super({
            direction: '',  // asc/desc/''
            width: 20,
            structure: {
                classes: ['sort-icon'],
                css: { float: 'right' }
            }
        }, ...args);
        this.direction = '';
        this.width = this.width || 20;

    }
    
    construct() {
        this.widget().css({
            width: this.width,
            marginLeft: this.width
        });
    }

    set_direction(direction) {
        if (direction !== this.direction) {
            this.direction = direction;
            this.notify('sort', direction);
        }
    }

    clear_sort() {
        this.set_direction('');
        this.draw();
    }

    toggle_sort_direction() {
        this.set_direction({
            asc: 'desc',
            desc: 'asc',
            '': 'desc'
        }[this.direction]);
        this.draw();
    }

    handlers() {
        var self = this;
        self.widget().on('click', function () {
            self.toggle_sort_direction();
        });
    }

    draw() {
        this.widget().empty();
        switch (this.direction) {
            case 'asc':
                this.widget().append(icon('up'));
                break;
            case 'desc':
                this.widget().append(icon('down'));
                break;
        }
    }
}
