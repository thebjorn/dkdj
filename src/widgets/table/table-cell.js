

import dk from "../../dk-obj";
import {Widget} from "../../widgetcore/dk-widget";


/*
 *  The TableCell class is responsible for pushing and pulling data into a
 *  table cell.  The data itself is contained in the enclosing TableRow's
 *  `record` property.
 */
export class TableCell extends Widget{
    constructor(...args) {
        super({
            tablerow: null,
            coldef: null,
        }, ...args);
    }

    set value(val) {
        if (!this.tablerow) return;
        this.tablerow.set_column_value(this.coldef, val);
    }
    get value () {
        if (!this.tablerow) return null;
        return this.tablerow.get_column_value(this.coldef);        
    }

    construct_edit_widget() {
        const location = this.widget();
        const coldef = this.coldef;
        const value = this.value;

        let width = location.width();
        location.empty();
        const edit_widget = coldef.widget_type.create_inside(location, {
            data: coldef.widget_data || {},  // eg. data to fill a drop-down list.
            url: coldef.widget_url,    // url to fetch data from
            css: { border: 'none' }
        });

        if (coldef.widget_type === 'SelectWidget') width += 20;  // XXX: yeuch!
        edit_widget.widget().width(width);
        edit_widget.value = value;  // does the right thing for select too..
        return edit_widget;
    }

    start_editing() {
        // prevent infinite recursion
        if (this.editing || this.coldef.editable === false) return;
        this.editing = true;

        if (!this._undo_stack) this._undo_stack = [];
        this._undo_stack.push(this.value);              // save current value, in case of undo.

        this.edit_widget = this.construct_edit_widget();

        // update field immediately on change (we can still undo..)
        this.edit_widget.widget().on('change', () => {
            const oldval = this.value;
            this.value = this.edit_widget.get_field_value();
            dk.trigger(this, 'dirty', this.coldef, this.value, this.edit_widget, oldval);
        });
    }

    save_editing() {
        if (this.coldef.editable === false) return;
        this.value = this.edit_widget.get_field_value();
        this._undo_stack.pop();
    }

    undo_editing() {
        if (this.coldef.editable === false) return;
        this.value = this._undo_stack.pop();
    }

    stop_editing() {
        if (this.coldef.editable === false) return;
        this._undo_stack = [];
        this.widget().empty();
        this.draw();
        this.editing = false;
    }

    draw(value) {
        const val = this.value === undefined ? value: this.value;
        this.widget().attr({
            axis: this.coldef.name,
            title: this.coldef.title
        }).html(this.coldef.format(val, this.tablerow.record, this.widget()));
    }
}
