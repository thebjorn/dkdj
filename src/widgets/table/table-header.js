

import dk from "../../dk-obj";
import {TableCell} from "./table-cell";
import {cursor} from "../cursors";
import {SortDirection} from "./sort-direction";

/*
 *  Column header widget.
 */
export class TableHeader  extends TableCell {
    constructor(...args) {
        super({
            scope: 'col',
            width: undefined,
            sortable: true,

            name: undefined,
            label: undefined,
            title: undefined,
            format: undefined,
            field: undefined,
            table: undefined,

            template: {root: 'th'},

        }, ...args);
    }
    

    construct() {
        if (this.sortable) {
            this.sort_icon = SortDirection.create_inside(this.widget(), {
                width: '2ex'
            });
        }
        this.title = this.description;
    }

    handlers() {
        const self = this;
        if (this.sortable) {
            self.widget().on('click', function () {
                self.table.set_sort(self);
            });
            dk.on(this.sort_icon, 'sort', () => this.set_sort_cursor());
        }
    }

    set_sort_cursor() {
        if (this.sortable) {
            const _cursor = this.sort_icon.direction === 'desc' ? 'down' : 'up';
            this.widget().css('cursor', cursor(_cursor));
            // this.table.notify('sort', this.sort_icon.direction);
            // this.table.trigger('sort', this.sort_icon.direction);
            dk.trigger(this.table, 'sort', this.sort_icon.direction);
        }
    }

    draw() {
        this.widget().attr({
            title: this.title,
            scope: this.scope,
            fieldname: this.name
        }).append(this.label);
        if (this.sortable) this.set_sort_cursor();
    }
}
