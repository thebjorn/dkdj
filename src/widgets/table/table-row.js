
/*
 *  A Data Record widget.
 */

import {Widget} from "../../widgetcore/dk-widget";
import {TableCell} from "./table-cell";
import {TableRowLayout} from "../../layout/table-layout";


export class TableRow extends Widget {
    constructor(...args) {
        super({
            dklayout: TableRowLayout,
            rownum: null,
            record: null,
            table: null,
            cells: [],
            template: {root: 'tr'},
        }, ...args);
    }
    
    delete_widget() {
        delete this.record;
        delete this.table;
        this.cells.forEach(c => c.delete_widget());
        this.cells = [];
        super.delete_widget();
    }

    construct() {
        const record = this.record;
        const table = this.table;
        this.widget().attr({
            pk: record.pk,
            rownum: this.rownum
        });
        this.cells = table.column_order.map(fname => {
            const coldef = table.column[fname];
            return TableCell.create_on(this.layout.add_td(), {
                tablerow: this,
                coldef: coldef
            });
        });
    }

    get_column_value(coldef) {
        return coldef.get_value(this.record);
    }
    set_column_value(coldef, val) {
        this.record[coldef.name] = val;
    }
}
