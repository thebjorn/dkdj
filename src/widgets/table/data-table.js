/*
 *  A DataTable is the normal table representation of a DataSource,
 *  i.e. records in rows, with each field in its own column (with a column
 *  header).
 */

import dk from "../../dk-obj";
import {TableLayout} from "../../layout/table-layout";
import {Widget} from "../../widgetcore/dk-widget";
import {TableRow} from "./table-row";
import is from "../../is";
import {ColumnDef} from "./column-def";
import {TableHeader} from "./table-header";
import {DataSet} from "../../data/dk-dataset";
import {AjaxDataSource} from "../../data/source/dk-ajax-datasource";
import {dkwarning} from "../../lifecycle/coldboot/dkwarning";


export class DataTableLayout extends TableLayout {
    constructor(widget, location, template, structure) {
        super(widget, location, template, structure);
        this.thead_tr = this.add_header_row();
    }
    
    add_thead_th(attrs) {
        const th = this.make('thead_th', attrs);
        this.thead_tr.appendln(th);
        return th;
    }
}


export class DataTable extends Widget {
    constructor(...args) {
        super({
            // type: 'datatable',
            dklayout: DataTableLayout,
            TableRow: TableRow,
            current_pagenum: 0,
            translate: {},  // l10n

            structure: {
                thead: {},
                tfoot: {},
                tbody: {}
            },

            template: {
                root: 'table',
                table: '<table/>',
                thead: '<thead/>',
                thead_tr: '<tr/>',
                thead_th: '<th/>',
                tbody: '<tbody/>',
                tbody_tr: '<tr/>'
            },

            table_data: null,
            state: undefined,

            // The columns attribute should be a property object that defines the visible columns:
            //
            //      columns: {
            //          colnam1: {}
            //          colname2: {..}
            //      ..
            //
            // this.columns is converted to an array of dk.table.ColumnDef, and this.column to a
            // dict from column name to dk.table.ColumnDef.
            columns: null,
            column: {},

            filters: {},

            // should table be sortable?
            sortable: true,

            // selector of dom element to bless as a download link (e.g.
            // <a id="foo" data-download="filename.csv">download</a>
            // =>  download: '#foo',
            download: undefined,

            header: null,
            rows: null,

            column_order: null,
            column_defs: null,            
        }, ...args);

        const self = this;
        this.rows = [];
        this.header = [];   // th cell objects

        this.column = {};
        this.column_order = [];
        if (!!this.datasource && this.table_data === null) {
            const dsprops = {datasource: this.datasource};
            if (this.pagesize != null) dsprops.pagesize = this.pagesize;
            if (this.orphans != null) dsprops.orphans = this.orphans;
            if (this.pagenum != null) dsprops.pagenum = this.pagenum;
            this.table_data = new DataSet(dsprops);
        }
        if (this.table_data === null) {
            this.data_url = (this.dk && this.dk.url)? this.dk.url : "";
            //dk.info("URL", this.data_url);
            // if data is not defined, then create one
            this.table_data = new DataSet({
                // default data source is AjaxDataSource
                datasource: new AjaxDataSource({
                    // default url is current url
                    url: self.data_url
                })
            });
        }
        if (!this.data_url) this.data_url = this.table_data.datasource.url;

        //        this.state = dk.State.create({
        //            state: this.table_data.current_pagedef,
        //            engine: dk.HashStorage.create({})
        //        });
        //this.state.restore();
    }

    construct() {
        // this.state = dk.page.hash.substate(this.id);

        if (this.download) {
            const $download = dk.$(this.download);
            const filename = $download.attr('data-download') || 'filename.csv';
            dk.$(this.download).attr('href', this.data_url + "!get-records?fmt=csv&filename=" + filename);
        }
    }

    /*
     *  Connect to a pager widget.
     */
    set_pager(pwidget) {
        // const self = this;
        if (this.table_data.page) {
            pwidget.set_pagecount(this.table_data.page.recordset.meta.pagecount);
            pwidget.select_page(this.table_data.page.query.pagenum);
        }

        dk.on(pwidget, 'select-page', (...args) => this.table_data.get_pagenum(...args));
        dk.on(this.table_data, 'fetch-info', (info, query) => {
            pwidget.set_pagecount(info.pagecount);
            pwidget.select_page(query.pagenum);
        });
    }

    filter_values() {
        return {};
    }

    handlers() {
        //$(this.table_data).on('fetch-data', _.bind(this.draw, this));
        //$bind('fetch-data@data -> draw@me', {data: this.table_data, me: this});
        dk.on(this.table_data, 'fetch-data-start', () => this.start_busy());
        dk.on(this.table_data, 'fetch-data', (...args) => this.draw(...args));
        dk.on(this.table_data, 'fetch-data', (...args) => this.end_busy(...args));
    }

    set_sort(sortcol) {
        if (!sortcol.sortable) return;
        if (this.sortable) {
            this.header.forEach(function (col) {
                if (col.sortable && col !== sortcol) col.sort_icon.clear_sort();
            });
            sortcol.sort_icon.toggle_sort_direction();
            this.table_data.set_sort(sortcol.name, sortcol.sort_icon.direction);
            dk.trigger(this, 'sort-order', sortcol.name, sortcol.sort_icon.direction, this);
        }
    }

    set_search(terms) {
        this.table_data.set_search(terms);
        if (this.state) this.state.set('pagedef', this.table_data.current_pagedef);
    }

    set_filter(vals) {
        this.table_data.set_filter(vals);
    }

    get_xy(colnum, rownum) {
        return this.rows[rownum].cells[colnum];
    }

    set_xy(colnum, rownum, val, html, title) {
        const cell = this.get_xy(colnum, rownum);
        cell.set_value(val, html, title);
        return cell;
    }

    values() {
        const res = [];
        this.rows.forEach(function (row) {
            const r = [];
            row.cells.forEach(function (cell) {
                r.push(cell.widget().text());
            });
            res.push(r);
        });
        return res;
    }

    /*
     *  Convert this.columns to an array of basic property objects.
     *
     *    {foo: {label: 'Foo', ..}} ==> [{name: 'foo', label: 'Foo', ...}, ...]
     */
    _get_columndefs(fieldlst) {
        const self = this;
        let columndefs;
        const fields = {};  // field.name -> field
        fieldlst.forEach(function (f) { fields[f.name] = f; });

        // make sure this.columns is an array of pojos
        if (is.isFunction(this.columns)) columndefs = this.columns(this.table_data.fields);
        else if (this.columns === null) columndefs = fieldlst;
        else if (!Array.isArray(this.columns)) {
            // the columns property has been defined, convert it to array
            columndefs = Object.keys(this.columns).map(name => {
                const col = this.columns[name];
                col.name = name;
                return col;
            });
        } else {
            dkwarning(`Expected {field1: {...}, field2: {...}, etc.}. columns property of unknown type: ${this.columns}`);
        }

        // convert the pojos to ColumnDef objects
        return columndefs.map(function (coldef, colnum) {
            const field = fields[coldef.name];
            const tmp = {};
            if (field) {
                Object.assign(tmp, {
                    fieldpos: field.pos,
                    fieldtype: field.type,
                    sortable: field.sortable,
                    label: field.label,
                    help_text: field.help_text,
                    field_widget: field.widget,
                    field_data: field.data
                });
            }
            Object.assign(tmp, coldef);
            Object.assign(tmp, {
                colnum: colnum,
                table: self
            });
            if (tmp.sortable === undefined) {
                tmp.sortable = self.sortable;
            }
            const cdef = new ColumnDef(tmp);
            if (fields[cdef.name]) cdef.bind_to_field(fields[cdef.name]);
            self.column[cdef.name] = cdef;
            self.column_order.push(cdef.name);
            return cdef;
        });
    }

    /*
     *  Draw the column headers.
     */
    draw_header(dataset) {  // must be a separate method so VDataTable can inherit..
        if (!this.drawn_header) {
            const self = this;
            // NOTE: we have data here!
            this.columns = this._get_columndefs(dataset.page.fields);

            this.columns.forEach(function (coldef) {
                const th = TableHeader.create(self.layout.add_thead_th(), coldef);
                self.header.push(th);
            });
            self.drawn_header = true;  // only draw header once.
        }
    }

    draw(dataset) {
        const self = this;

        if (!dataset) {   // NOTE: data source consumer pattern..?  (on: data -> draw?)
            this.table_data.get_page({
                pagenum: this.current_pagenum
            });
        } else {
            self.draw_header(dataset);
            dk.trigger(self, 'draw-start', self);
            self.layout.clear_body();
            self.rows = [];

            dataset.page.records.forEach(function (record, rownum) {
                const tr = self.TableRow.create(self.layout.add_row_to('tbody'), {
                    rownum: rownum,
                    record: record,
                    table: self
                });
                self.rows.push(tr);
            });
            dk.trigger(self, 'draw-end', self);
        }
    }
}
