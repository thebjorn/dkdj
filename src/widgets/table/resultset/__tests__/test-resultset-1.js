
import $ from 'jquery';
import {PagerWidget} from "../../../pager-widget";
import page from "../../../../widgetcore/dk-page";
import {DataSet} from "../../../../data/dk-dataset";
import {DataTable} from "../../../table/data-table";
import {ArraySource} from "../../../../data/source/dk-array-datasource";
import {ResultSet} from "../resultset";
import utidy from "../../../../browser/dk-html";


test("test-pagerwidget-with-datatable-datasource-shortcut", () => {
    document.body.innerHTML = `
        <div id="work">
        </div>
    `;
    const work = $('#work');
    page.initialize(document);

    ResultSet.create_inside(work, {
        construct_table: function (location, downloadwidget) {
            return DataTable.create_on(location, {
                classes: ['table table-bordered table-hover table-condensed'],
                download: downloadwidget,
                datasource: [
                    {project: 'first', work: '1:03:57'},
                    {project: 'aTiktok', work: '2:44:57'},
                    {project: 'bGenerelt NT', work: '1:03:57'},
                    {project: 'cTiktok', work: '2:44:57'},
                    {project: 'dGenerelt NT', work: '1:03:57'},
                    {project: 'eTiktok', work: '2:44:57'},
                    {project: 'fGenerelt NT', work: '1:03:57'},
                    {project: 'gTiktok', work: '2:44:57'},
                    {project: 'hGenerelt NT', work: '1:03:57'},
                    {project: 'iTiktok', work: '2:44:57'},
                    {project: 'jGenerelt NT', work: '1:03:57'},
                    {project: 'kTiktok', work: '2:44:57'},
                    {project: 'last', work: '1:06:43'}
                ],
                pagesize: 5,
                orphans: 4,
                columns: {
                    project: {label: 'Prosjekt'},
                    work: {label: 'Arbeid'}
                }
            });
        }
    });
    
    // console.log(utidy(work.html()));

    expect(document.getElementById('work')).toMatchSnapshot();
});
