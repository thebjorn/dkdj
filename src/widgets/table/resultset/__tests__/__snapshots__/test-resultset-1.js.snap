// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-pagerwidget-with-datatable-datasource-shortcut 1`] = `
<div
  id="work"
>
  
        
  <section
    class="ResultSet clearfix"
    id="result-set-1"
  >
    <div
      class="rowbx"
      id="dk-bx-1"
      style="display: flex; align-items: flex-start;"
    >
      <div
        class="filterbx"
        id="dk-bx-2"
        style="margin-right: 0px; width: 0px; padding-right: 0px;"
      />
      <div
        class="panel panel-default data"
        id="dk-bx-3"
        style="flex: 1px;"
      >
        <header
          class="panel-heading clearfix"
          id="dk-bx-4"
        >
          <div
            class="pull-left search"
            id="dk-bx-5"
            style="width: 40%;"
          >
            <form
              class="SearchWidget search-form form-inline"
              id="search-widget-1"
            >
              <div
                class="form-group has-feedback"
                id="dk-bx-6"
              >
                <div
                  class="input-group input-group-sm"
                  id="dk-bx-7"
                >
                  <input
                    accesskey="s"
                    class="form-control search-input"
                    id="dk-bx-8"
                    name="q"
                    placeholder="Søk.."
                    type="search"
                  />
                  <span
                    class="clear-button form-control-feedback"
                    id="dk-bx-9"
                    style="display: none;"
                  >
                    <i
                      class="icon fa fa-times times fa-fw"
                      icon="times"
                      value="times:fw"
                    />
                  </span>
                  <div
                    class="input-group-btn go-group"
                    id="dk-bx-10"
                  >
                    <button
                      class="go-button btn btn-primary"
                      id="dk-bx-11"
                    >
                      <i
                        class="icon fa fa-search search fa-fw"
                        icon="search"
                        value="search:fw"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </form>
          </div>
          <div
            class="commands pull-right"
            id="dk-bx-12"
          >
            <img
              alt="add"
              class="add-button add"
              height="28"
              id="dk-bx-13"
              src="//static.datakortet.no/ikn/ax/button_blue_add.png"
            />
            <button
              class="btn excelbtn csv"
              data-loading-text="Laster ned..."
              id="dk-bx-14"
            >
              
                            
              <span>
                <img
                  alt="csv"
                  height="28"
                  src="//static.datakortet.no/ikn/ax/button_blue_down.png"
                />
              </span>
              
                        
            </button>
          </div>
          <div
            class="pull-right stats"
            id="dk-bx-15"
            style="text-align: right; margin-right: 1ex;"
          />
        </header>
        <table
          class="result-table table table-bordered table-hover table-condensed content DataTable dk-bx"
          id="dk-bx-16"
        >
          <thead
            id="dk-bx-18"
          >
            

            <tr
              id="dk-table-1"
            >
              

              <th
                class="string"
                fieldname="project"
                id="dk-table-2"
                scope="col"
                style="cursor: url(//static.datakortet.no/dk/sort-za.cur), pointer;"
                title=""
              >
                <div
                  class="SortDirection sort-icon"
                  id="sort-direction-1"
                  style="float: right; width: 2ex; margin-left: 2ex;"
                />
                Prosjekt
              </th>
              

              

              <th
                class="string"
                fieldname="work"
                id="dk-table-3"
                scope="col"
                style="cursor: url(//static.datakortet.no/dk/sort-za.cur), pointer;"
                title=""
              >
                <div
                  class="SortDirection sort-icon"
                  id="sort-direction-2"
                  style="float: right; width: 2ex; margin-left: 2ex;"
                />
                Arbeid
              </th>
              

            </tr>
            

          </thead>
          <tfoot
            id="dk-bx-19"
          />
          <tbody
            id="dk-bx-20"
          >
            

            <tr
              class="TableRow"
              id="dk-table-4"
              pk="0"
              rownum="0"
            >
              

              <td
                axis="project"
                class="TableCell"
                id="dk-tr-1"
              >
                first
              </td>
              

              

              <td
                axis="work"
                class="TableCell"
                id="dk-tr-2"
              >
                1:03:57
              </td>
              

            </tr>
            

            

            <tr
              class="TableRow"
              id="dk-table-5"
              pk="1"
              rownum="1"
            >
              

              <td
                axis="project"
                class="TableCell"
                id="dk-tr-3"
              >
                aTiktok
              </td>
              

              

              <td
                axis="work"
                class="TableCell"
                id="dk-tr-4"
              >
                2:44:57
              </td>
              

            </tr>
            

            

            <tr
              class="TableRow"
              id="dk-table-6"
              pk="2"
              rownum="2"
            >
              

              <td
                axis="project"
                class="TableCell"
                id="dk-tr-5"
              >
                bGenerelt NT
              </td>
              

              

              <td
                axis="work"
                class="TableCell"
                id="dk-tr-6"
              >
                1:03:57
              </td>
              

            </tr>
            

            

            <tr
              class="TableRow"
              id="dk-table-7"
              pk="3"
              rownum="3"
            >
              

              <td
                axis="project"
                class="TableCell"
                id="dk-tr-7"
              >
                cTiktok
              </td>
              

              

              <td
                axis="work"
                class="TableCell"
                id="dk-tr-8"
              >
                2:44:57
              </td>
              

            </tr>
            

            

            <tr
              class="TableRow"
              id="dk-table-8"
              pk="4"
              rownum="4"
            >
              

              <td
                axis="project"
                class="TableCell"
                id="dk-tr-9"
              >
                dGenerelt NT
              </td>
              

              

              <td
                axis="work"
                class="TableCell"
                id="dk-tr-10"
              >
                1:03:57
              </td>
              

            </tr>
            

          </tbody>
        </table>
        <footer
          class="panel-footer"
          id="dk-bx-17"
        >
          <ul
            class="PagerWidget pagination"
            id="pager-widget-1"
            style="margin-top: 0px; margin-bottom: 0px;"
          >
            <li
              class="prev-nav"
            >
              <a
                href="#"
              >
                «
              </a>
            </li>
            <li
              class="page active"
              page="1"
            >
              <a
                href="#"
              >
                1
              </a>
            </li>
            <li
              class="page"
              page="2"
            >
              <a
                href="#"
              >
                2
              </a>
            </li>
            <li
              class="next-nav"
            >
              <a
                href="#"
              >
                »
              </a>
            </li>
          </ul>
        </footer>
      </div>
    </div>
  </section>
</div>
`;
